modal/_asgi.py

```diff
@@ -167,3 +170,194 @@ def webhook_asgi_app(fn: Callable, method: str):
+        )[20:24]
+    )
+
+
+def wait_for_web_server(host: str, port: int, *, timeout: float) -> None:
+    """Wait until a web server port starts accepting TCP connections."""
+    import socket
+    import time
+
+    start_time = time.monotonic()
+    while True:
+        try:
+            with socket.create_connection((host, port), timeout=timeout):
+                break
+        except OSError as ex:
+            time.sleep(0.01)
+            if time.monotonic() - start_time >= timeout:
+                raise TimeoutError(
+                    f"Waited too long for port {port} to start accepting connections. "
+                    + "Make sure the web server is listening on all interfaces, or adjust `startup_timeout`."
+                ) from ex
+
+
+async def _proxy_http_request(session: aiohttp.ClientSession, scope, receive, send) -> None:
+    proxy_response: Optional[aiohttp.ClientResponse] = None
+
+    async def request_generator():
+        while True:
+            message = await receive()
+            if message["type"] == "http.request":
+                body = message.get("body", b"")
+                if body:
+                    yield body
+                if not message.get("more_body", False):
+                    break
+            elif message["type"] == "http.disconnect":
+                proxy_response.connection.transport.abort()  # Abort the connection.
+                break
+            else:
+                raise ExecutionError(f"Unexpected message type: {message['type']}")
+
+    path = scope["path"]
+    if scope.get("query_string"):
+        path += scope["query_string"].decode()
+
+    proxy_response = await session.request(
+        method=scope["method"],
+        url=path,
+        headers=[(k.decode(), v.decode()) for k, v in scope["headers"]],
+        data=request_generator(),
+        allow_redirects=False,
+    )
+
+    async def send_response():
+        msg = {
+            "type": "http.response.start",
+            "status": proxy_response.status,
+            "headers": [(k.encode(), v.encode()) for k, v in proxy_response.headers.items()],
+        }
+        await send(msg)
+        async for data in proxy_response.content.iter_any():
+            msg = {
+                "type": "http.response.body",
+                "body": data,
+                "more_body": True,
+            }
+            await send(msg)
+        await send({"type": "http.response.body"})
+
+    async def listen_for_disconnect():
+        while True:
+            message = await receive()
+            if message["type"] == "http.disconnect":
+                proxy_response.connection.transport.abort()
+
+    try:
+        send_response_task = asyncio.create_task(send_response())
+        disconnect_task = asyncio.create_task(listen_for_disconnect())
+        await asyncio.wait([send_response_task, disconnect_task], return_when=asyncio.FIRST_COMPLETED)
+    finally:
+        send_response_task.cancel()
```

Comment:
In case any of the spawned tasks raise exceptions, it could be good to log them (or reraise since it probably means something is broken?) before cancelling the task, otherwise they get lost and never surface.
E.g.
```
if send_response_task.done():
    send_response_task.result()   # raises the exception from the task if there was one
```